File: src/kodeks/tools/registry.py
"""Deterministic local tool definitions for Python runtime parity."""

from __future__ import annotations

import json
import os
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from ..storage import KodeksDatabase
from ..workspace import (
    ShellCommandTimeoutError,
    WorkspaceService,
    run_command,
)
from .schemas import ToolDefinition, default_tool_definitions

ToolArguments = Mapping[str, Any]
ToolExecutionStatus = str


@dataclass(frozen=True)
class ToolExecutionContext:
    """Carries session and tool-call ids into approval and audit records."""

    session_id: str | None = None
    tool_call_id: str | None = None


@dataclass(frozen=True)
class ToolExecutionResult:
    """Model-facing tool execution result."""

    status: ToolExecutionStatus
    output: str


@dataclass(frozen=True)
class ToolRegistryServices:
    """Service bundle used by deterministic local tool handlers."""

    workspace: WorkspaceService
    database: KodeksDatabase
    environment: Mapping[str, str | None] | None = None


@dataclass(frozen=True)
class RegisteredTool:
    """One registered tool definition and its Python handler."""

    definition: ToolDefinition
    read_only: bool
    mutating: bool
    handler: Callable[[ToolArguments, ToolExecutionContext], ToolExecutionResult]


class ToolRegistry:
    """Stores deterministic tool definitions and handlers by public tool name."""

    def __init__(self, tools: list[RegisteredTool]) -> None:
        self._tools = {tool.definition["name"]: tool for tool in tools}

    def definitions(self, read_only_only: bool = False) -> list[ToolDefinition]:
        """Return provider-facing definitions in stable registration order."""

        return [
            tool.definition
            for tool in self._tools.values()
            if not read_only_only or (tool.read_only and not tool.mutating)
        ]

    def has(self, tool_name: str) -> bool:
        """Return whether a tool name is registered."""

        return tool_name in self._tools

    def execute(
        self,
        tool_name: str,
        arguments: ToolArguments,
        context: ToolExecutionContext | None = None,
    ) -> ToolExecutionResult:
        """Execute one registered tool and convert unknown tools into failures."""

        tool = self._tools.get(tool_name)
        if tool is None:
            return failed_output(f"Unknown tool: {tool_name}")
        return tool.handler(arguments, context or ToolExecutionContext())


def build_default_tool_registry(services: ToolRegistryServices) -> ToolRegistry:
    """Build the default Kodeks tool registry from local services."""

    definitions = default_tool_definitions()
    return ToolRegistry(
        [
            RegisteredTool(
                definitions[0],
                True,
                False,
                lambda arguments, context: execute_read_file(arguments, services),
            ),
            RegisteredTool(
                definitions[1],
                False,
                True,
                lambda arguments, context: execute_write_file(arguments, services),
            ),
            RegisteredTool(
                definitions[2],
                True,
                False,
                lambda arguments, context: execute_grep(arguments, services),
            ),
            RegisteredTool(
                definitions[3],
                False,
                True,
                lambda arguments, context: execute_run_shell(
                    arguments, context, services
                ),
            ),
            RegisteredTool(
                definitions[4],
                False,
                True,
                lambda arguments, context: execute_remember_fact(
                    arguments, context, services
                ),
            ),
            RegisteredTool(
                definitions[5],
                True,
                False,
                lambda arguments, context: execute_recall_memory(arguments, services),
            ),
            RegisteredTool(
                definitions[6],
                True,
                False,
                lambda arguments, context: execute_read_memory_artifact(
                    arguments, services
                ),
            ),
            RegisteredTool(
                definitions[7],
                True,
                False,
                lambda arguments, context: execute_spawn_explore_agent(
                    arguments, context, services
                ),
            ),
            RegisteredTool(
                definitions[8],
                True,
                False,
                lambda arguments, context: execute_list_mcp_servers(services),
            ),
            RegisteredTool(
                definitions[9],
                True,
                False,
                lambda arguments, context: execute_list_skills(arguments, services),
            ),
            RegisteredTool(
                definitions[10],
                True,
                False,
                lambda arguments, context: execute_read_skill(arguments, services),
            ),
        ]
    )


def execute_read_file(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """Execute read_file through the shared workspace boundary."""

    path = string_argument(arguments, "path")
    if path is None:
        return failed_output("read_file requires a non-empty string path")
    try:
        return completed_output(
            {"ok": True, "path": path, "content": services.workspace.read_file(path)}
        )
    except Exception as exc:
        return failed_output(error_message(exc), {"path": path})


def execute_write_file(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """Execute write_file with whole-file overwrite semantics."""

    path = string_argument(arguments, "path")
    content = string_argument(arguments, "content", allow_empty=True)
    if path is None:
        return failed_output("write_file requires a non-empty string path")
    if content is None:
        return failed_output("write_file requires string content", {"path": path})
    try:
        services.workspace.write_file(path, content)
        return completed_output(
            {
                "ok": True,
                "path": path,
                "strategy": "whole_file_overwrite",
                "bytesWritten": len(content.encode()),
            }
        )
    except Exception as exc:
        return failed_output(error_message(exc), {"path": path})


def execute_grep(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """Execute a literal grep over visible workspace text files."""

    query = string_argument(arguments, "query")
    limit = clamp_integer(arguments.get("limit"), 1, 1000, 20)
    if query is None:
        return failed_output("grep requires a non-empty string query")
    matches: list[dict[str, Any]] = []
    for path in services.workspace.list_files():
        if len(matches) >= limit:
            break
        try:
            content = services.workspace.read_file(path)
        except Exception:
            continue
        for index, line in enumerate(content.splitlines(), start=1):
            if query in line:
                matches.append({"path": path, "line": index, "text": line})
            if len(matches) >= limit:
                break
    return completed_output({"ok": True, "query": query, "matches": matches})


def execute_run_shell(
    arguments: ToolArguments,
    context: ToolExecutionContext,
    services: ToolRegistryServices,
) -> ToolExecutionResult:
    """Execute run_shell and record approval requests for dangerous commands."""

    command = string_argument(arguments, "command")
    if command is None:
        return failed_output("run_shell requires a non-empty string command")
    try:
        result = run_command(command, services.workspace.root_path())
    except ShellCommandTimeoutError:
        return failed_output("Command timed out")
    except Exception as exc:
        return failed_output(error_message(exc))
    if result.approval_required:
        approval = services.database.approvals.create_approval(
            command={"command": command},
            reason=result.stderr,
            session_id=context.session_id,
            tool_call_id=context.tool_call_id,
        )
        services.database.audit_log.record(
            context.session_id,
            "approval_required",
            {"approvalId": approval.id, "command": command},
        )
        return ToolExecutionResult(
            "approval_required",
            json_output(
                {
                    "ok": False,
                    "approvalRequired": True,
                    "approvalId": approval.id,
                    "status": approval.status,
                    "reason": approval.reason,
                    "command": command,
                }
            ),
        )
    return completed_output(
        {
            "ok": result.exit_code == 0,
            "command": command,
            "exitCode": result.exit_code,
            "stdout": result.stdout,
            "stderr": result.stderr,
            "approvalRequired": False,
        }
    )


def execute_remember_fact(
    arguments: ToolArguments,
    context: ToolExecutionContext,
    services: ToolRegistryServices,
) -> ToolExecutionResult:
    """Execute remember_fact through the memory repository."""

    content = string_argument(arguments, "content")
    scope = string_argument(arguments, "scope") or "project"
    if content is None:
        return failed_output("remember_fact requires non-empty string content")
    memory_id = services.database.memories.remember(
        scope, content, context.session_id
    )
    return completed_output(
        {"ok": True, "memoryId": memory_id, "scope": scope, "content": content}
    )


def execute_recall_memory(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """Execute recall_memory through the memory repository."""

    query = string_argument(arguments, "query")
    limit = clamp_integer(arguments.get("limit"), 1, 50, 5)
    if query is None:
        return failed_output("recall_memory requires a non-empty string query")
    layers = read_memory_layers(arguments.get("layers"))
    return completed_output(
        {
            "ok": True,
            "query": query,
            "layers": layers,
            "memories": services.database.memories.recall(query, limit),
            "layered": services.database.memories.recall_layered(
                query, limit, layers
            ),
        }
    )


def execute_read_memory_artifact(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """Execute read_memory_artifact through the memory repository."""

    ref_id = string_argument(arguments, "refId")
    if ref_id is None:
        return failed_output("read_memory_artifact requires a non-empty string refId")
    artifact = services.database.memories.read_artifact_content(ref_id)
    if artifact is None:
        return failed_output(f"Unknown memory artifact: {ref_id}", {"refId": ref_id})
    return completed_output({"ok": True, "refId": ref_id, **artifact})


def execute_spawn_explore_agent(
    arguments: ToolArguments,
    context: ToolExecutionContext,
    services: ToolRegistryServices,
) -> ToolExecutionResult:
    """Execute a minimal read-only explore subagent run."""

    task = string_argument(arguments, "task")
    if task is None:
        return failed_output("spawn_explore_agent requires a non-empty string task")
    run = services.database.subagents.start_run(
        context.session_id or "session_unknown", "explore", task
    )
    summary = f"Explore agent completed task: {task}"
    completed = services.database.subagents.complete_run(str(run["id"]), summary)
    return completed_output(
        {
            "ok": True,
            "runId": completed["id"],
            "status": completed["status"],
            "summary": completed["summary"],
        }
    )


def execute_list_mcp_servers(services: ToolRegistryServices) -> ToolExecutionResult:
    """List configured MCP server manifests without opening network clients."""

    servers = read_mcp_server_manifests(runtime_environment(services))
    return completed_output({"ok": True, "servers": servers, "count": len(servers)})


def execute_list_skills(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """List available skill directories and titles."""

    query = string_argument(arguments, "query")
    limit = clamp_integer(arguments.get("limit"), 1, 50, 20)
    skills = discover_skills(services)
    if query is not None:
        lowered = query.lower()
        skills = [
            skill
            for skill in skills
            if lowered in f"{skill['name']}\n{skill['title']}".lower()
        ]
    return completed_output({"ok": True, "skills": skills[:limit]})


def execute_read_skill(
    arguments: ToolArguments, services: ToolRegistryServices
) -> ToolExecutionResult:
    """Read one discovered skill body by exact directory name."""

    name = string_argument(arguments, "name")
    if name is None:
        return failed_output("read_skill requires a non-empty string name")
    for skill in discover_skills(services):
        if skill["name"] == name:
            path = Path(str(skill["path"]))
            return completed_output(
                {
                    "ok": True,
                    "name": skill["name"],
                    "title": skill["title"],
                    "content": path.read_text(),
                }
            )
    return failed_output(f"Unknown skill: {name}")


def string_argument(
    arguments: ToolArguments, name: str, allow_empty: bool = False
) -> str | None:
    """Read a required string argument from model-provided JSON."""

    value = arguments.get(name)
    if not isinstance(value, str):
        return None
    trimmed = value.strip()
    if not allow_empty and not trimmed:
        return None
    return value if allow_empty else trimmed


def clamp_integer(value: object, minimum: int, maximum: int, fallback: int) -> int:
    """Read and clamp a numeric argument from model-provided JSON."""

    if not isinstance(value, int) or isinstance(value, bool):
        return fallback
    return min(maximum, max(minimum, value))


def read_memory_layers(value: object) -> list[str]:
    """Read optional memory layer filters from model-provided JSON."""

    if not isinstance(value, list):
        return ["atom", "scenario", "artifact"]
    layers = [
        item
        for item in value
        if item in {"atom", "scenario", "artifact"} and isinstance(item, str)
    ]
    return layers or ["atom", "scenario", "artifact"]


def read_mcp_server_manifests(
    environment: Mapping[str, str | None],
) -> list[dict[str, Any]]:
    """Read MCP server manifests from environment variables."""

    raw_servers = environment.get("KODEKS_MCP_SERVERS")
    if raw_servers:
        return parse_mcp_server_manifests(raw_servers)
    url = environment.get("KODEKS_MCP_SERVER_URL")
    if not url:
        return []
    return [
        {
            "label": environment.get("KODEKS_MCP_SERVER_LABEL") or "default",
            "url": url.strip(),
            "allowedTools": split_csv(environment.get("KODEKS_MCP_ALLOWED_TOOLS")),
            "skipApproval": environment.get("KODEKS_MCP_SKIP_APPROVAL") == "true",
        }
    ]


def parse_mcp_server_manifests(raw_servers: str) -> list[dict[str, Any]]:
    """Parse JSON MCP manifests while discarding malformed entries."""

    try:
        parsed = json.loads(raw_servers)
    except json.JSONDecodeError:
        return []
    items = parsed if isinstance(parsed, list) else [parsed]
    manifests: list[dict[str, Any]] = []
    for item in items:
        if not isinstance(item, dict):
            continue
        label = item.get("label")
        url = item.get("url")
        if not isinstance(label, str) or not isinstance(url, str):
            continue
        raw_allowed = item.get("allowedTools")
        allowed_tools = (
            [tool for tool in raw_allowed if isinstance(tool, str)]
            if isinstance(raw_allowed, list)
            else split_csv(raw_allowed if isinstance(raw_allowed, str) else None)
        )
        manifests.append(
            {
                "label": label,
                "url": url,
                "allowedTools": allowed_tools,
                "skipApproval": item.get("skipApproval") is True,
            }
        )
    return manifests


def split_csv(value: str | None) -> list[str]:
    """Split comma-separated environment values into non-empty tokens."""

    if value is None:
        return []
    return [item.strip() for item in value.split(",") if item.strip()]


def discover_skills(services: ToolRegistryServices) -> list[dict[str, str]]:
    """Discover SKILL.md files from configured roots."""

    skills: list[dict[str, str]] = []
    for root in skill_roots(services):
        if not root.is_dir():
            continue
        for child in sorted(root.iterdir(), key=lambda item: item.name):
            skill_path = child / "SKILL.md"
            if not child.is_dir() or not skill_path.is_file():
                continue
            content = skill_path.read_text()
            skills.append(
                {
                    "name": child.name,
                    "title": read_markdown_title(content) or child.name,
                    "path": str(skill_path),
                }
            )
    return skills


def skill_roots(services: ToolRegistryServices) -> list[Path]:
    """Resolve skill roots from env or the workspace-local default."""

    environment = runtime_environment(services)
    configured = split_csv(environment.get("KODEKS_SKILLS_PATHS"))
    roots = (
        configured
        if configured
        else [str(Path(services.workspace.root_path()) / ".kodeks" / "skills")]
    )
    return [Path(os.path.expanduser(root)).resolve() for root in roots]


def read_markdown_title(content: str) -> str | None:
    """Read the first markdown heading as a skill title."""

    for line in content.splitlines():
        if line.startswith("# "):
            return line[2:].strip()
    return None


def completed_output(payload: dict[str, Any]) -> ToolExecutionResult:
    """Create a successful JSON tool result."""

    return ToolExecutionResult("completed", json_output(payload))


def failed_output(
    message: str, extra: dict[str, Any] | None = None
) -> ToolExecutionResult:
    """Create a failed JSON tool result."""

    return ToolExecutionResult(
        "failed", json_output({"ok": False, "error": message, **(extra or {})})
    )


def json_output(payload: dict[str, Any]) -> str:
    """Serialize tool outputs compactly for model-facing tool messages."""

    return json.dumps(payload, separators=(",", ":"))


def runtime_environment(services: ToolRegistryServices) -> Mapping[str, str | None]:
    """Return configured tool environment or process env."""

    return services.environment if services.environment is not None else os.environ


def error_message(error: object) -> str:
    """Convert unknown thrown values into readable tool errors."""

    return str(error)

File: frontend/lib/server/tools/registry.ts
// frontend/lib/server/tools/registry.ts
// ToolRegistry + 9 个 handler：逐字节忠实移植 Python src/kodeks/tools/registry.py。
// 输出形状逐字段；审批是第三种工具状态；handler 异步（M2 存储异步）。
//
// 保真红线（见 50-tools-security.md §1/§4、保真风险 1/2/8/9/10）：
//  · execute 未知工具 → failed_output("Unknown tool: <name>")。
//  · run_shell 危险命令分支产 status:'approval_required'，输出键序 ok,approvalRequired,approvalId,status,reason,command；
//    并 createApproval({command}) + auditLog.record(sid,'approval_required',{approvalId,command})。
//  · read_memory_artifact 输出键序 ok,refId,artifact,content（artifact 字段展开）。
//  · spawn_explore_agent 是确定性桩：session||'session_unknown'、listFiles(12)、evidence=visible[:5]、CERCN 逐字。
//  · plan-mode schema 由 definitions(readOnlyOnly) 裁剪；执行层 allowlist 守卫在 agent/tool-loop.ts。
//  · mutating/read_only 只驱动 mode allowlist，不替代 run_shell 审批。
import {
  type ShellResult,
  ShellCommandTimeoutError,
  runCommand,
} from '../workspace'
import {
  clampInteger,
  completedOutput,
  errorMessage,
  failedOutput,
  jsonOutput,
  readMcpServerManifests,
  readMemoryLayers,
  runtimeEnvironment,
  stringArgument,
} from './helpers'
import { toolDefinitionsByName } from './schemas'
import type {
  RegisteredTool,
  ToolArguments,
  ToolDefinition,
  ToolExecutionContext,
  ToolExecutionResult,
  ToolRegistryServices,
} from './types'

/** 按公共工具名存确定性工具定义与 handler（移植 ToolRegistry，registry.py:32-63）。 */
export class ToolRegistry {
  private readonly tools: Map<string, RegisteredTool>

  /** @param tools 已注册工具列表，按注册顺序入 Map（Map 保留插入序）。 */
  constructor(tools: RegisteredTool[]) {
    this.tools = new Map(tools.map((tool) => [tool.definition.name, tool]))
  }

  /**
   * 按稳定注册顺序返回 provider-facing 定义（移植 definitions，registry.py:38-45）。
   * readOnlyOnly=true 时过滤条件为 tool.readOnly && !tool.mutating（与 schema 裁剪结果一致，独立路径）。
   */
  definitions(readOnlyOnly = false): ToolDefinition[] {
    const result: ToolDefinition[] = []
    for (const tool of this.tools.values()) {
      if (!readOnlyOnly || (tool.readOnly && !tool.mutating)) {
        result.push(tool.definition)
      }
    }
    return result
  }

  /** 返回工具名是否已注册（移植 has，registry.py:47-50）。 */
  has(toolName: string): boolean {
    return this.tools.has(toolName)
  }

  /**
   * 执行一个已注册工具，未知工具转为失败（移植 execute，registry.py:52-63）。
   * @param context 可选执行上下文；缺省为空 context（sessionId/toolCallId 均 undefined）。
   */
  async execute(
    toolName: string,
    args: ToolArguments,
    context: ToolExecutionContext = {},
  ): Promise<ToolExecutionResult> {
    const tool = this.tools.get(toolName)
    if (tool === undefined) {
      return failedOutput(`Unknown tool: ${toolName}`)
    }
    return tool.handler(args, context)
  }
}

/**
 * 从本地服务构建默认 Kodeks 工具注册表（移植 build_default_tool_registry，registry.py:66-135）。
 * 9 个工具按固定顺序注册，readOnly/mutating 标志逐字对齐 Python。
 */
export function buildDefaultToolRegistry(services: ToolRegistryServices): ToolRegistry {
  const definitions = toolDefinitionsByName()
  return new ToolRegistry([
    {
      definition: definitions.read_file,
      readOnly: true,
      mutating: false,
      handler: (args) => executeReadFile(args, services),
    },
    {
      definition: definitions.write_file,
      readOnly: false,
      mutating: true,
      handler: (args) => executeWriteFile(args, services),
    },
    {
      definition: definitions.grep,
      readOnly: true,
      mutating: false,
      handler: (args) => executeGrep(args, services),
    },
    {
      definition: definitions.run_shell,
      readOnly: false,
      mutating: true,
      handler: (args, context) => executeRunShell(args, context, services),
    },
    {
      definition: definitions.remember_fact,
      readOnly: false,
      mutating: true,
      handler: (args, context) => executeRememberFact(args, context, services),
    },
    {
      definition: definitions.recall_memory,
      readOnly: true,
      mutating: false,
      handler: (args) => executeRecallMemory(args, services),
    },
    {
      definition: definitions.read_memory_artifact,
      readOnly: true,
      mutating: false,
      handler: (args) => executeReadMemoryArtifact(args, services),
    },
    {
      definition: definitions.spawn_explore_agent,
      readOnly: true,
      mutating: false,
      handler: (args, context) => executeSpawnExploreAgent(args, context, services),
    },
    {
      definition: definitions.list_mcp_servers,
      readOnly: true,
      mutating: false,
      handler: () => executeListMcpServers(services),
    },
  ])
}

/** 经共享 workspace 边界执行 read_file（移植 execute_read_file，registry.py:138-151）。 */
async function executeReadFile(
  args: ToolArguments,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const path = stringArgument(args, 'path')
  if (path === null) {
    return failedOutput('read_file requires a non-empty string path')
  }
  try {
    return completedOutput({
      ok: true,
      path,
      content: services.workspace.readFile(path),
    })
  } catch (error) {
    return failedOutput(errorMessage(error), { path })
  }
}

/** 用整文件覆盖语义执行 write_file（移植 execute_write_file，registry.py:154-176）。 */
async function executeWriteFile(
  args: ToolArguments,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const path = stringArgument(args, 'path')
  // content 用 allowEmpty=true：允许空串且返回原始未 trim 值。
  const content = stringArgument(args, 'content', true)
  if (path === null) {
    return failedOutput('write_file requires a non-empty string path')
  }
  if (content === null) {
    return failedOutput('write_file requires string content', { path })
  }
  try {
    services.workspace.writeFile(path, content)
    return completedOutput({
      ok: true,
      path,
      strategy: 'whole_file_overwrite',
      // bytesWritten 用 UTF-8 字节长度（不是 content.length），对齐 Python len(content.encode())。
      bytesWritten: Buffer.byteLength(content, 'utf8'),
    })
  } catch (error) {
    return failedOutput(errorMessage(error), { path })
  }
}

/** 在可见工作区文本文件上做字面 grep（移植 execute_grep，registry.py:179-201）。 */
async function executeGrep(
  args: ToolArguments,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const query = stringArgument(args, 'query')
  const limit = clampInteger(args.limit, 1, 1000, 20)
  if (query === null) {
    return failedOutput('grep requires a non-empty string query')
  }
  const matches: Array<{ path: string; line: number; text: string }> = []
  for (const path of services.workspace.listFiles()) {
    if (matches.length >= limit) {
      break
    }
    let content: string
    try {
      content = services.workspace.readFile(path)
    } catch {
      continue
    }
    // splitlines 等价：按行切分且不含换行符；行号从 1 起。
    const lines = content.split(/\r\n|\r|\n/)
    // Python splitlines 不会为末尾换行产生空尾行；这里若内容以换行结尾则去掉末尾空串。
    if (lines.length > 0 && lines[lines.length - 1] === '' && /[\r\n]$/.test(content)) {
      lines.pop()
    }
    let index = 0
    for (const line of lines) {
      index += 1
      if (line.includes(query)) {
        matches.push({ path, line: index, text: line })
      }
      if (matches.length >= limit) {
        break
      }
    }
  }
  return completedOutput({ ok: true, query, matches })
}

/** 执行 run_shell 并为危险命令记录审批请求（移植 execute_run_shell，registry.py:204-254）。 */
async function executeRunShell(
  args: ToolArguments,
  context: ToolExecutionContext,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const command = stringArgument(args, 'command')
  if (command === null) {
    return failedOutput('run_shell requires a non-empty string command')
  }
  let result: ShellResult
  try {
    result = await runCommand(command, services.workspace.rootPath())
  } catch (error) {
    if (error instanceof ShellCommandTimeoutError) {
      return failedOutput('Command timed out')
    }
    return failedOutput(errorMessage(error))
  }
  if (result.approvalRequired) {
    // 审批路径：创建 pending 审批 + 审计 approval_required，返回 approval_required 状态。
    const approval = await services.database.approvals.createApproval(
      { command },
      result.stderr,
      context.sessionId ?? null,
      context.toolCallId ?? null,
    )
    await services.database.auditLog.record(
      context.sessionId ?? null,
      'approval_required',
      { approvalId: approval.id, command },
    )
    return {
      status: 'approval_required',
      // 键序逐字：ok,approvalRequired,approvalId,status,reason,command。
      output: jsonOutput({
        ok: false,
        approvalRequired: true,
        approvalId: approval.id,
        status: approval.status,
        reason: approval.reason,
        command,
      }),
    }
  }
  // 正常路径：ok = exitCode === 0（timeout/parse 失败走异常/审批分支，不进此处）。
  return completedOutput({
    ok: result.exitCode === 0,
    command,
    exitCode: result.exitCode,
    stdout: result.stdout,
    stderr: result.stderr,
    approvalRequired: false,
  })
}

/** 经内存 repository 执行 remember_fact（移植 execute_remember_fact，registry.py:257-273）。 */
async function executeRememberFact(
  args: ToolArguments,
  context: ToolExecutionContext,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const content = stringArgument(args, 'content')
  // scope 默认 "project"（Python `string_argument(...) or "project"`，空串也回退用 ||）。
  const scope = stringArgument(args, 'scope') || 'project'
  if (content === null) {
    return failedOutput('remember_fact requires non-empty string content')
  }
  const memoryId = await services.database.memories.remember(
    scope,
    content,
    context.sessionId ?? null,
  )
  return completedOutput({ ok: true, memoryId, scope, content })
}

/** 经内存 repository 执行 recall_memory（移植 execute_recall_memory，registry.py:276-296）。 */
async function executeRecallMemory(
  args: ToolArguments,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const query = stringArgument(args, 'query')
  const limit = clampInteger(args.limit, 1, 50, 5)
  if (query === null) {
    return failedOutput('recall_memory requires a non-empty string query')
  }
  const layers = readMemoryLayers(args.layers)
  return completedOutput({
    ok: true,
    query,
    layers,
    memories: await services.database.memories.recall(query, limit),
    layered: await services.database.memories.recallLayered(query, limit, layers),
  })
}

/** 经内存 repository 执行 read_memory_artifact（移植 execute_read_memory_artifact，registry.py:299-310）。 */
async function executeReadMemoryArtifact(
  args: ToolArguments,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const refId = stringArgument(args, 'refId')
  if (refId === null) {
    return failedOutput('read_memory_artifact requires a non-empty string refId')
  }
  const artifact = await services.database.memories.readArtifactContent(refId)
  if (artifact === null) {
    return failedOutput(`Unknown memory artifact: ${refId}`, { refId })
  }
  // 键序逐字：ok,refId,artifact,content（artifact dict 的 artifact/content 字段展开）。
  return completedOutput({ ok: true, refId, ...artifact })
}

/** 执行一个最小只读 explore 子代理 run（移植 execute_spawn_explore_agent，registry.py:313-368）。 */
async function executeSpawnExploreAgent(
  args: ToolArguments,
  context: ToolExecutionContext,
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const task = stringArgument(args, 'task')
  if (task === null) {
    return failedOutput('spawn_explore_agent requires a non-empty string task')
  }
  // 缺省字面量 "session_unknown"（Python `context.session_id or "session_unknown"`）。
  const sessionId = context.sessionId || 'session_unknown'
  const run = await services.database.subagents.startRun(sessionId, 'explore', task)
  // 最多 12 个文件。
  const visibleFiles = services.workspace.listFiles(12)
  // 逐字顺序固定的四项 allowedTools。
  const allowedTools = ['read_file', 'grep', 'recall_memory', 'read_memory_artifact']
  await services.database.auditLog.record(sessionId, 'subagent_started', {
    runId: run.id,
    agentName: run.agentName,
    task,
    allowedTools,
    visibleFileCount: visibleFiles.length,
  })
  const summary = subagentSummary(task, visibleFiles)
  const contract = subagentContract(task, visibleFiles)
  const completed = await services.database.subagents.completeRun(run.id, summary)
  await services.database.auditLog.record(sessionId, 'subagent_completed', {
    runId: completed.id,
    agentName: completed.agentName,
    status: completed.status,
    summary: completed.summary,
    allowedTools,
    contract,
  })
  return completedOutput({
    ok: true,
    runId: completed.id,
    status: completed.status,
    summary: completed.summary,
    allowedTools,
    parentSessionId: sessionId,
    contract,
    quarantine: {
      readOnly: true,
      canMutateWorkspace: false,
      canRequestApproval: false,
    },
  })
}

/** 不开网络客户端地列出已配置 MCP server manifests（移植 execute_list_mcp_servers，registry.py:371-375）。 */
async function executeListMcpServers(
  services: ToolRegistryServices,
): Promise<ToolExecutionResult> {
  const servers = readMcpServerManifests(runtimeEnvironment(services))
  return completedOutput({ ok: true, servers, count: servers.length })
}

/**
 * 从工作区清单构建有界只读子代理摘要（移植 _subagent_summary，registry.py:378-385）。
 * preview = visible[:5] 用 ", " 连接；无文件时 "no visible files"。文案逐字。
 */
function subagentSummary(task: string, visibleFiles: string[]): string {
  const preview =
    visibleFiles.length > 0 ? visibleFiles.slice(0, 5).join(', ') : 'no visible files'
  return `Read-only explore run completed. Task: ${task}. Visible workspace sample: ${preview}.`
}

/**
 * 为子代理输出构建结构化、便于综合的契约（移植 _subagent_contract，registry.py:388-398）。
 * evidence = visible[:5]；confidence 有 evidence → "medium" 否则 "low"。CERCN 字段顺序逐字。
 */
function subagentContract(
  task: string,
  visibleFiles: string[],
): {
  claim: string
  evidence: string[]
  risk: string
  confidence: string
  nextAction: string
} {
  const evidence = visibleFiles.slice(0, 5)
  return {
    claim: 'Read-only workspace exploration completed.',
    evidence,
    risk: 'This run only sampled visible files and did not mutate or execute commands.',
    confidence: evidence.length > 0 ? 'medium' : 'low',
    nextAction: `Main agent should synthesize this with the parent task: ${task}.`,
  }
}
